mlir/lib/Dialect/IRDL/IRDLConstraint.cpp

```diff
@@ -0,0 +1,245 @@
+//===- IRDLConstraint.cpp - IRDL constraints definition ---------*- C++ -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This file declares the different type constraints an operand or a result can
+// have.
+//
+//===----------------------------------------------------------------------===//
+
+#include "mlir/Dialect/IRDL/IRDLConstraint.h"
+#include "mlir/Dialect/IRDL/IR/IRDL.h"
+
+using namespace mlir;
+using namespace irdl;
+
+//===----------------------------------------------------------------------===//
+// Explicit instanciations
+//===----------------------------------------------------------------------===//
+
+namespace mlir {
+namespace irdl {
+
+template class EqConstraint<Type>;
```

Comment:
Hmm, I think that this wouldn't work that way.
When you include `IRDLConstraint.h`, you need that either the code is there, or a specialization that indicates that the code is somewhere else (i.e. here).
Here, all methods that use `template <Item>` need to me moved in `IRDLConstraint.h`, and the one with `template <>` should probably be declared in `IRDLConstraint.h`, and keep their definition here.
So the list you have here should not be needed, but should be moved in the `.h`

Reply:
See previous comment.